### ISSUE TITLE
Lag while drawing

### ISSUE THREAD

**Issue by tincancrab** [drawpile/Drawpile#575]
I'm getting a lot of increased complaints with lag with each update of Drawpile. While I myself do not experience severe lag except on rare occasions (which I have noticed a bit more often with updates), some experience it all the time. These complaints also mainly stem from heavy traffic rooms. If there is a way to either optimize Drawpile for large sessions and/or help find the main source of this problem, that would be great.

I realize the simplest fix would be a server-wide setting to limit max user count per room, but I'd like to see this only as a last resort, due to this being the amount of users drawing versus lurking and the nature of Drawpile being a communal canvas. Limiting this would make users feel forced to find alternative servers or programs.

Through this post, I'm going to try and have multiple users post the details on their lag problems, the version they are using, and their computer specs to find the main culprit behind this increasing problem (though I'm afraid this may be optimization, as previously stated). Seems to have spiked with the release of 2.0.6.

**Comment by callaa** [comment]
There haven't been any changes lately that I can think of that would make the application more laggy, so I suspect the problem is related to increased traffic, rather than regressions in the application itself.

The root cause, I think, is that Drawpile wasn't originally designed for huge rooms with lots of users coming and going, so the architecture is not optimal for that use case. That said, I believe there are still things I can do to improve the performance in general.

Here are some performace tips:

 * A large canvas will eventually get slow for two reasons: it requires a large cache buffer and, more importantly, it eventually fills with stuff. An empty canvas, no matter how large, consumes very little memory.
 * Layers are cheap, but layer content is not. A layer that is mostly transparent consumes very little memory and has hardly any effect on session load time. If you have multiple layers with drawings that do not overlap, it's almost as cheap and fast as a single layer.
 * Try to use the largest brush spacing % you can get away with. A large brush with 1% spacing kills the CPU.
 * Watercolor brush is slow. The pixel (hard edged) brush is the most efficient, but not much more than the regular soft brush. (If I add additional brush engines in 2.1, I'll add special optimizations for the pixel brush.)
 * The direct drawing mode is slightly more efficient than the indirect mode.
 * Pasting/stamping fills up the session history fast. There is no way around that.
 * Moving and rescaling selections without cutting&pasting is light on the bandwidth, because no external pixel data needs to be transmitted.
 * Use a session size limit that's appropriate for the canvas size. (I'm going to make this adjustable per session in 2.1) A rule of thumb is that the size limit should be at least twice the size of the session saved as an ORA. Otherwise, resets will occur much too frequently, since the reset image itself eats up the lion's share of the session history.
 * Don't touch the background layer. Drawpile has a special optimization for layers filled entirely with a solid color. Even a single pixel of some other color breaks this optimization. (I will fix this in version 2.1)
 * Drawpile only refreshes the part of the canvas that is visible. Using the Navigator dock breaks this optimization, since the whole canvas is always visible.

I have a few ideas I'm going to implement in 2.1 to improve performance, but efficient support for large scale sessions with lots of users drawing simultaneously is a hard problem. A radical rearchitecturing would fix some problems (e.g. a server that has it's own copy of the paint engine would remove the need for resets, but would also introduce new problems; chief of which is would be increased system requirements. The public server could maybe host one or two simultaneous sessions.)

Finally, a request regarding lag problem reports: Please try to use some other word to describe the problem than "lag". This is purely to help me understand the problem better. The word "lag" is quite ambigous, or at least people are using it in a variety of nonstandard ways. In my mind, lag is network latency or, in some contexts, the delay between user action and visual feedback. However, I've seen people use the word to refer to all sorts of unrelated things, including features they do not like or understand or any bug they're affected by.

**Comment by SpaceDustCM** [comment]
Thanks for the note man! So in laymans' terms, it just seems to be program limitations? Would make sense to me, as a person who uses PS and Clip (for animation--whew does performance sputter when animation files get BIG).

I suppose a better term for the later of what I was referring to with log loading slowly is that it's almost like each layer is " printed", so to speak. Even at times when all pieces on board are drawn ( we try to limit the amount of pasted work as possible ), and the reset image be 5mb or below, it does the occasional "printing". 
I have to test when it actually does this what it's like (time-people#-MB). The timeframes I recorded last night were about 7-10 seconds each reset with about 45 people in the room and the log being between 4-4.7MB. For reference, the public room has 9 layers ( this includes the solid Background layer ).

**Comment by tincancrab** [comment]
Lag, in my case, seems to be the program not detecting all pen input (for example, I write "so how are you doing today?" on the canvas but "so h   a  u doing tod   " is what happens on the canvas).

**Comment by callaa** [comment]
@SpaceDustCM The printing effect you're seeing (the image appearing in tiles?) is normal and expected: the reset is a snapshot of the rasterized canvas. The idea is that drawing the canvas as an image is faster than replaying the entire history. For this reason, it's critical that the history size limit is set correctly: too low and resets happen too often or worse, the reset image is bigger than the size limit. I have an idea for an optimization I can do in version 2.1.0 that will make this a bit faster.

@tincancrab That sounds like mouse event compression. Do you have tablet bug workaround mode enabled?

**Comment by tincancrab** [comment]
Yes, it's enabled.

**Comment by callaa** [comment]
@tincancrab If you're using version 2.0.9, can you try disabling the bug workaround mode, or does that break something for you? When the workaround mode is enabled, mouse events are used instead of tablet events, which means queued events get dropped when the system is under load.

**Comment by askmeaboutlo0m** [comment]
I think this should be resolved with Drawpile 2.2. It's gotten significant performance improvements in regards to pixel pushing, it offloads painting to a separate thread and it renders the canvas in the background, which means the navigator no longer totally obliterates performance.

If there are still performance problems, please open new issues for those.
